mergify_engine/worker.py

```diff
@@ -0,0 +1,332 @@
+            else:
+                LOG.error(
+                    "fail to process stream after %s attemps",
+                    attemps,
+                    gh_repo=repo,
+                    gh_owner=owner,
+                    gh_pull=pull_number,
+                    exc_info=True,
+                )
+
+    score = utils.utcnow().timestamp()
+    await redis.eval(
+        ATOMIC_CLEAN_STREAM_SCRIPT,
+        keys=[stream_name],
+        args=[score, len(message_ids)] + message_ids,
+    )
+
+
+class Worker:
+    def __init__(self, worker_count=config.STREAM_WORKERS, idle_sleep_time=0.42):
+        self._idle_sleep_time = idle_sleep_time
+        self._worker_count = worker_count
+        self._loop = asyncio.get_running_loop()
+        self._tombstone = asyncio.Event()
+        self._workers_running = asyncio.Event()
+        self._pusher_running = asyncio.Event()
+        self._pusher_tombstone = asyncio.Event()
+        self._worker_tasks = []
+
+        self._queue = asyncio.Queue(maxsize=self._worker_count)
+        self._executor = concurrent.futures.ThreadPoolExecutor(
+            max_workers=self._worker_count
+        )
+
+        self._installation_locks = collections.defaultdict(asyncio.Lock)
+        self._redis = None
+
+        self.pending_streams = set()
+
+    async def stream_worker_task(self, worker_id):
+        # NOTE(sileht): This task must never fail, we don't want write code to reap/clean/respawn them
+        while self._workers_running.is_set():
+            try:
+                try:
+                    stream_name = await asyncio.wait_for(
+                        self._queue.get(), timeout=self._idle_sleep_time
+                    )
+                except asyncio.TimeoutError:
+                    continue
+
+                LOG.debug("stream worker %d take stream: %s", worker_id, stream_name)
+                self.pending_streams.add(stream_name)
+
+                try:
+                    installation_id = stream_name.split("~")[1]
+                    async with self._installation_locks[installation_id]:
+                        async with get_sources_from_stream(
+                            self._redis, stream_name
+                        ) as sources:
+                            await self._loop.run_in_executor(
+                                self._executor,
+                                functools.partial(run_engine, stream_name, sources),
+                            )
+                finally:
+                    LOG.debug(
+                        "stream worker %d release stream: %s", worker_id, stream_name
+                    )
+                    self.pending_streams.remove(stream_name)
+                    self._queue.task_done()
+            except Exception:
+                LOG.error("stream worker %d fail", worker_id, exc_info=True)
+                return
+
+        LOG.info("stream worker %d exited", worker_id)
+
+    async def push_streams_to_worker(self):
```

Comment:
e.g having this in its own class might make the Worker easier to understand/follow

Reply:
I completly remove this method.